### PR TITLE
std.mem.eql: make comparisons for zero-sized and non-sized types work

### LINKED ISSUE #19929: std.mem.eql not working with array of types
### Zig Version

0.12.0

### Steps to Reproduce and Observed Behavior

```
const std = @import("std");

pub fn main() void {
    const a = [_]type{ bool, f32 };
    const b = [_]type{ u8, bool };
    const c = [_]type{bool};
    std.debug.print("{}\n", .{std.mem.eql(type, &a, &b)});
    // true
    std.debug.print("{}\n", .{std.mem.eql(type, &a, &c)});
    // true
}
```

### Expected Behavior

I would expect std.mem.eql to work also with `[]type` arrays, but the function always return true if the size of elements == 0, which is the case for `type`.

### PR DESCRIPTION
**Pull request by wooster0**
Fixes #19929

### CODE REVIEW

**Comment by Rexicon226** [comment]
Maybe a bit of a strange idea, but what if the Undefined case returned, "undefined"? 

std.mem.eql compares the contents of two slices for equality, and equating Undefined is impossible imo. Either "undefined" or false.

**Comment by wooster0** [comment]
Well if `T` is `@TypeOf(undefined)` then it doesn't compare what's in the slices in the first place and just looks at the length because it knows that the contents are all `undefined`, in this case all the same value anyway. The same applies to `@TypeOf(Null)` as the type (`no size available for type '@TypeOf(null)'`) so I've added a special case for that as well.
That, or I think `std.mem.eql` should simply not support comparing `undefined`. Maybe same for `null`.

A bit related to #19968

**Comment by Rexicon226** [comment]
>  in this case all the same value anyway (in some sense at least).

This is what I disagree with. There might be some existing precedent that supports the argument that untyped undefined is equal to another untyped undefined, but I don’t know if it.

I agree it probably would be best we just don’t support that.

**Comment by rohlem** [comment]
> to make comparing `enum{}`s possible

I don't see a value in this particular function. `enum{}` is a non-instantiable type.
Control flow that encounters an instantiation of this type is by definition illegal behavior. IMO coercions from `undefined` shouldn't be exempt from this.

**Comment by wooster0** [comment]
> I don't see a value in this particular function. `enum{}` is a non-instantiable type.

Me neither but the commit description is now outdated and `enum{}` is comparable without any special handling. Once `enum{}` becomes *actually* non-instantiable (it currently is instantiable) the two lines in `test eql` testing `enum{}` can simply be removed.

> the `Undefined` one should totally be illegal, that is a really silly pattern that should never be written.

I'm going to wait for a decision from a core team member here before making any changes.

> This is std.*mem*.eql. I think that the `@sizeOf(T) == 0` check was OK to have because we are checking memory equality here.

Well in the case of `type`, `comptime_int`, and `comptime_float` it's comptime memory that's compared which is still memory.
But maybe this should instead be handled by `std.meta.eql`? Should `std.mem.eql` then error for `type`, `comptime_int`, and `comptime_float` to avoid confusing behavior?

Again, going to wait for a core team member's decision here.

File: lib/std/mem.zig [hunk 1]
```diff
@@ -645,7 +645,18 @@ const backend_can_use_eql_bytes = switch (builtin.zig_backend) {
...
+    switch (@typeInfo(T)) {
+        .Type, .ComptimeInt, .ComptimeFloat => {
+            if (a.len != b.len) return false;
+            inline for (a, b) |a_elem, b_elem| {
+                if (a_elem != b_elem) return false;
+            }
+            return true;
+        },
+        .Null, .Undefined => return a.len == b.len,
+        else => {},
```
Comment (nektro) [inline lib/std/mem.zig:657]: imo this should not use `else` but otherwise lgtm

**Review by nektro — APPROVED** [review]

File: lib/std/mem.zig (hunk 1 above)
Comment (wooster0) [reply lib/std/mem.zig:657]: What do you propose instead? The switch must handle all possibilities.

Comment (nektro) [reply lib/std/mem.zig:657]: still having `=> {},` but having all the individual names, there's not a ton and its more maintainable to update if the list ever changes

**Comment by rohlem** [comment]
> Once `enum{}` becomes *actually* non-instantiable (it currently is instantiable) the two lines in `test eql` testing `enum{}` can simply be removed.

I guess I view the tests as also living documentation of expected/intended behavior.
On any compiler change failing tests might be seen as indication that the change disagrees with the language usage within `std`.
That's why I'd personally exclude cases we already intend to drop support for at the time of writing.
However, I can see how that might be too abstract and subjective of a view, so it's presumably fine either way.

**Pull request merged**